test/tokengenerator.go

```diff
@@ -0,0 +1,141 @@
+package testutils
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+	"github.com/lestrrat-go/jwx/jwk"
+)
+
+const (
+	bitSize = 2048
+)
+
+// WebKeySet represents a JWK Set object.
+type WebKeySet struct {
+	Keys []jwk.Key `json:"keys"`
+}
+
+// ExtraClaim a function to set claims in the token to generate
+type ExtraClaim func(token *jwt.Token)
+
+// WithEmailClaim sets the `email` claim in the token to generate
+func WithEmailClaim(email string) ExtraClaim {
+	return func(token *jwt.Token) {
+		token.Claims.(jwt.MapClaims)["email"] = email
+	}
+}
+
+// Identity is a user identity
+type Identity struct {
+	ID       uuid.UUID
+	Username string
+}
+
+// NewIdentity returns a new, random identity
+func NewIdentity() *Identity {
+	return &Identity{
+		ID:       uuid.NewV4(),
+		Username: "testuser-" + uuid.NewV4().String(),
+	}
+}
+
+// TokenGenerator represents the test token and key manager.
+type TokenGenerator struct {
+	keyMap map[string]*rsa.PrivateKey
+}
+
+// NewTokenGenerator creates a new TokenGenerator.
+func NewTokenGenerator() *TokenGenerator {
+	tg := &TokenGenerator{}
+	tg.keyMap = make(map[string]*rsa.PrivateKey)
+	return tg
+}
+
+// CreateKey creates and stores a new key with the given kid.
+func (tg *TokenGenerator) CreateKey(kid string) (*rsa.PrivateKey, error) {
+	reader := rand.Reader
+	key, err := rsa.GenerateKey(reader, bitSize)
+	if err != nil {
+		return nil, err
+	}
+	tg.keyMap[kid] = key
+	return key, nil
+}
+
+// Key retrieves the key associated with the given kid.
+func (tg *TokenGenerator) Key(kid string) (*rsa.PrivateKey, error) {
+	key, ok := tg.keyMap[kid]
+	if !ok {
+		return nil, errors.New("given kid does not exist")
+	}
+	return key, nil
+}
+
+// GenerateSignedToken generates a JWT user token and signs it using the default private key
+func (tg *TokenGenerator) GenerateSignedToken(identity Identity, kid string, extraClaims ...ExtraClaim) (string, error) {
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Claims.(jwt.MapClaims)["uuid"] = identity.ID
+	token.Claims.(jwt.MapClaims)["preferred_username"] = identity.Username
+	token.Claims.(jwt.MapClaims)["sub"] = identity.ID
+	token.Claims.(jwt.MapClaims)["jti"] = uuid.NewV4().String()
+	token.Claims.(jwt.MapClaims)["session_state"] = uuid.NewV4().String()
+	token.Claims.(jwt.MapClaims)["iat"] = time.Now().Unix()
+	token.Claims.(jwt.MapClaims)["exp"] = time.Now().Unix() + 60*60*24*30
+	token.Claims.(jwt.MapClaims)["nbf"] = 0
+	token.Claims.(jwt.MapClaims)["iss"] = "fabric8-auth"
+	token.Claims.(jwt.MapClaims)["typ"] = "Bearer"
+	token.Claims.(jwt.MapClaims)["approved"] = true
```

Comment:
We won't have this "approved" claim in our prod tokens. It's fabric8-auth specific thing. It doesn't harm though.